packages/wallet/frontend/src/middleware.ts

```diff
@@ -52,9 +51,13 @@ export async function middleware(req: NextRequest) {
     }
   } else {
     // If the user is not logged in and tries to access a private resource,
-    // redirect to auth page.
+    // redirect to auth page or in the case of grant-interaction, back to the interaction page.
+    const callbackUrl =
+      req.url.indexOf('grant-interact') !== -1 ? `?callbackUrl=${req.url}` : ''
     if (!isPublic && !response.success) {
-      return NextResponse.redirect(new URL('/auth', req.url))
+      return NextResponse.redirect(
+        new URL(`/auth/login/${callbackUrl}`, req.url)
+      )
```

Comment:
Instead of doing it only for the interaction page we should do it for every private routes.